src/Moq/Protected/ProtectedMock.cs

```diff
@@ -359,6 +383,193 @@ private static Expression<Action<T>> GetSetterExpression(PropertyInfo property,
+		{
+			var properties = typeof(T).GetProperties(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
+			var matchingProperties = properties.Where(p => p.Name == propertyName).ToList();
+			if (matchingProperties.Count > 1)
+			{
+				return GetMatchingIndexer(matchingProperties, CastValueForIndexer(value, propertyName), isSetter, exact);
+			}
+			if (matchingProperties.Count == 0)
+			{
+				return null;
+			}
+			return matchingProperties[0];
+		}
+
+		private static PropertyInfo GetMatchingIndexer(List<PropertyInfo> indexers, object[] keys, bool isSetter, bool exact)
+		{
+			var argTypes = ToArgTypes(keys);
+
+			var matchingIndexer = indexers.FirstOrDefault(p =>
+			{
+				if (isSetter)
+				{
+					if (!p.CanWrite) { return false; }
+					return p.GetSetMethod(true).GetParameterTypes().CompareTo(argTypes, exact, considerTypeMatchers: false);
+				}
+				else
+				{
+					if (!p.CanRead) { return false; }
+					return p.GetGetMethod(true).GetParameterTypes().CompareTo(argTypes, exact, considerTypeMatchers: false);
+				}
+
+			});
+
+			return matchingIndexer;
+		}
+		
+		private static Expression<Action<T>> GetIndexerSetterExpression(string indexerName, object[] indexerKeys, object value)
+		{
+			Guard.NotNullOrEmpty(indexerName, nameof(indexerName));
+			Guard.NotNullOrEmpty(indexerKeys, nameof(indexerKeys));
+
+			var setupSetArguments = indexerKeys.Concat(new object[] { value }).ToArray();
+			var indexer = GetPropertyIncludingIndexer(indexerName, false, indexerKeys, false);
+			ThrowIfMemberMissing(indexerName, indexer);
+			Guard.IsIndexer(indexer, nameof(indexerName));
+			ThrowIfSetterNotApplicable(indexer);
+
+			return GetIndexerSetterExpression(indexer, setupSetArguments);
+		}
+
+		private static Expression<Action<T>> GetIndexerSetterExpression(PropertyInfo property, object[] indexerKeysAndValue)
+		{
+			var param = Expression.Parameter(typeof(T), "mock");
+
+			return Expression.Lambda<Action<T>>(
+				Expression.Call(param, property.GetSetMethod(true), IndexerArgsToExpressions(indexerKeysAndValue, property, true)),
+				param);
+		}
+
+		private static LambdaExpression GetIndexerGetterExpression(string indexerName, object[] indexerKeys)
+		{
+			Guard.NotNullOrEmpty(indexerName, nameof(indexerName));
+			Guard.NotNullOrEmpty(indexerKeys, nameof(indexerKeys));
+
+			var indexer = GetPropertyIncludingIndexer(indexerName, false, indexerKeys, false);
+			ThrowIfMemberMissing(indexerName, indexer);
+			Guard.IsIndexer(indexer, nameof(indexerName));
+			ThrowIfGetterNotApplicable(indexer);
+
+			return GetIndexerGetterExpression(indexer, indexerKeys);
+		}
+
+		private static LambdaExpression GetIndexerGetterExpression(PropertyInfo property, object[] indexerKeys)
+		{
+			var param = Expression.Parameter(typeof(T), "mock");
+
+			return Expression.Lambda(
+				Expression.Call(param, property.GetGetMethod(true), IndexerArgsToExpressions(indexerKeys, property, true)),
```

Comment:
Shouldn't the last argument to `IndexerArgsToExpressions` (for its `isSet` parameter) be `false`?

Reply:
Yes that is an error.  Tests pass because the indexerKeys omit the property type.
Will correct.